juju/juju
# [Dev] record agent start time in machine document

## Description of change

When deploying to a provider that uses shadow IPs (eg. AWS), the public IP addresses are not exposed as network interfaces to the instance (so the machiner cannot pick them up) and can only be detected by querying the provider directly inside the instance poller.

The instance poller implementation in 2.6.x and 2.7-rcX (note that 2.7 has a new instancepoller implementation) partitions monitored machines into two groups: a short poll group that gets frequently polled and a long poll group that gets polled approximately every 10-15min. Once a machine is running and has at least one address assigned it gets migrated to the long poll group.

On EC2 (and other providers), there are times when the machine needs to be restarted.
When this happens, the machine will probably (unless it's assigned an elastic IP) get a new public IP. Once the agent reconnects to the API, the machine will show up as working again; however, the controller thinks that the machine still uses the **old public IP**. As a result:

- we cannot `juju ssh` in the machine;
- `juju upgrade-controller` seems to block until the new IP is detected.
 
The problem here is that unless we somehow notify the instancepoller to check the machine, we will need to wait ~10-15min until the next refresh interval!

This PR attempts to resolve this issue by:
- Adding a `AgentStartedAt` (agent-started-at in bson) field in the machine document and arranging for the machine agent to update it when it boots.
- Creating a new watcher that monitors the `[Life, AgentStartedAt]` tuple from the machine documents (ignoring containers) for changes. The instancepoller now uses this instead of the misleadingly-named `WatchModelMachines` which only reports **life-cycle** related changes. 

Notes:
- The new watcher tries to be conservative regarding the number of DB accesses by collecting changes and processing them in bulk after not seeing new changes for some time (configurable; currently set to 500ms). For more details see the extended log commit in https://github.com/juju/juju/pull/10902/commits/6da33b0aae89a0f4bd562e63cf722a3211be246e.
- The changes are split in a way so they can be back-ported to 2.6 **without** the recently introduced 2.7 changes to the instancepoller.
- In the future, we could potentially extend this notification mechanism to also include the link layer devices in the monitored tuple and implement some form of hot-plugging: machiner detects (inotify) a new eth device, link layer devices get updated, instance poller gets triggered and pulls the public (shadow) IP.

## QA steps

I have tested this on AWS but it should work with any provider that uses shadow IPs

```
$ juju bootstrap aws aws --credential aws --no-gui --build-agent
$ juju model-config logging-config='<root>=ERROR;juju.worker.instancepoller=DEBUG'
$ juju deploy cs:~jameinel/ubuntu-lite-7

# Tail the model logs and wait until the new machine enters the long poll group
$ juju debug-log --include-module juju.worker.instancepoller --replay --level DEBUG

# You should see something like this:
controller-0: 13:03:49 INFO juju.worker.instancepoller machine "0" (instance ID "i-05fe715dc5a6e6c6c") instance status changed from {"allocating" "Start instance attempt 1"} to {"running" "running"}
controller-0: 13:03:49 INFO juju.worker.instancepoller machine "0" (instance ID "i-05fe715dc5a6e6c6c") has new addresses: [public:34.228.219.51 local-cloud:172.31.46.138]
controller-0: 13:05:14 DEBUG juju.worker.instancepoller moving machine "0" (instance ID "i-05fe715dc5a6e6c6c") to long poll group

# Then (still tailing the logs) wait until the application shows up as running in `juju status`.
# Log in to the aws console, right click on the "juju-default-machine-0" instance, select
# "instance state" -> "stop".
# 
# Wait for the machine to shutdown and for juju status to show something like
# "agent lost".
#
# Then go back to the aes console and start the machine. After a couple of seconds you 
# should see something like this in the model logs:
controller-0: 13:07:49 DEBUG juju.worker.instancepoller moving machine "0" (instance ID "i-05fe715dc5a6e6c6c") to short poll group
controller-0: 13:07:53 INFO juju.worker.instancepoller machine "0" (instance ID "i-05fe715dc5a6e6c6c") has new addresses: [public:34.207.175.62 local-cloud:172.31.46.138]
controller-0: 13:07:53 DEBUG juju.worker.instancepoller moving machine "0" (instance ID "i-05fe715dc5a6e6c6c") to long poll group

# Double-check that `juju status` now lists the new public IP
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1852164
